build/resolver.go

```diff
@@ -0,0 +1,360 @@
+package build
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/containerd/containerd/platforms"
+	"github.com/docker/buildx/builder"
+	"github.com/docker/buildx/driver"
+	"github.com/docker/buildx/util/progress"
+	"github.com/moby/buildkit/client"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/util/flightcontrol"
+	"github.com/moby/buildkit/util/tracing"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+type resolvedNode struct {
+	resolver    *nodeResolver
+	driverIndex int
+	platforms   []specs.Platform
+}
+
+func (dp resolvedNode) Node() builder.Node {
+	return dp.resolver.nodes[dp.driverIndex]
+}
+
+func (dp resolvedNode) Client() *client.Client {
+	clients, err := dp.resolver.boot(context.TODO(), []int{dp.driverIndex}, nil)
+	if err != nil {
+		panic(err)
+	}
+	return clients[0]
+}
+
+func (dp resolvedNode) Opts() gateway.BuildOpts {
+	opts, err := dp.resolver.opts(context.TODO(), []int{dp.driverIndex}, nil)
+	if err != nil {
+		panic(err)
+	}
+	return opts[0]
+}
+
+type matcher struct {
+	platforms.Matcher
+	perfect bool
+}
+
+type nodeResolver struct {
+	nodes    []builder.Node
+	clients  flightcontrol.Group[*client.Client]
+	opt      flightcontrol.Group[gateway.BuildOpts]
+	matchers []func(specs.Platform) matcher
+}
+
+func newDriverResolver(nodes []builder.Node) *nodeResolver {
+	// create a collection of ordered matchers for each platform
+	var matcherCache sync.Map
```

Comment:
I'm not sure if this cache really buys us anything. The matchers are just value comparisons that are faster than map lookup (and I don't think there is a big chance of duplicates).